fix #862: make HIBP callback check email_addresses

groovecoder opened the pull request.

groovecoder commented on controllers/hibp.js:
```diff
@@ -13,6 +13,20 @@ const mozlog = require("../log");
 
+// Get address and language from either subscribers or
+// email_addresses fields
+function getAddressAndLanguageForEmail(recipient) {
```
There's probably a more stream-lined way to do this?

groovecoder replied:
Though maybe not, because we still want/need to add a "Send all breach alerts to my primary inbox" feature, and that logic is best added here.

groovecoder commented on tests/controllers/hibp.test.js:
```diff
@@ -72,6 +72,40 @@ test("notify POST with breach, subscriber hash prefix and suffixes should call s
 });
 
 
+test("good notify POST with breach, secondary email hash prefix and suffixes should call sendEmail and respond with 200", async () => {
```
lots of code duplication in this test, but I want to get it up for review before I refactor it.

groovecoder commented on controllers/hibp.js:
```diff
@@ -76,14 +89,14 @@ async function notify (req, res) {
           SERVER_URL: req.app.locals.SERVER_URL,
           scanAnotherEmailHref: scanAnotherEmailHref,
-          unsubscribeUrl: EmailUtils.getUnsubscribeUrl(subscriber, utmID),
+          unsubscribeUrl: EmailUtils.getUnsubscribeUrl(recipient, utmID),
```
TODO: need to test this.

pdehaan commented on controllers/hibp.js:
```diff
@@ -13,6 +13,20 @@ const mozlog = require("../log");
+// Get address and language from either subscribers or
+// email_addresses fields
+function getAddressAndLanguageForEmail(recipient) {
+  let email = "";
```
Not sure if we can do `let email = recipient.email;` and skip the `else` statement below.

groovecoder replied:
Yup, we can.

pdehaan commented on tests/controllers/hibp.test.js:
```diff
@@ -72,6 +72,40 @@ test("notify POST with breach, subscriber hash prefix and suffixes should call s
+  expect (mockFluentFormatCalls.length).toBe(1);
+  const mockFluentFormatCallArgs = mockFluentFormatCalls[0];
+  expect (mockFluentFormatCallArgs[0]).toEqual(["en"]);
+  expect (mockFluentFormatCallArgs[1]).toBe("hibp-notify-email-subject");
```
minor nit, but not sure if it'd be easier or more readable if we could destructure the `mockFLuentFormatCallArgs[]` to make these more readable.

```js
const [mockFluentLocales, mockFluentSubject] = mockFluentFormatCalls[0];
expect(mockFluentLocales).toEqual(["en"]);
expect(mockFluentSubject).toBe("hibp-notify-email-subject");
```

pdehaan commented on controllers/hibp.js:
```diff
@@ -45,27 +59,26 @@ async function notify (req, res) {
 
+  for (const recipient of recipients) {
+    log.info("notify", {recipient});
+    const { email, signupLanguage } = getAddressAndLanguageForEmail(recipient)
```
```
/home/travis/build/mozilla/blurts-server/controllers/hibp.js
  72:79  error  Missing semicolon  semi
```

groovecoder commented:
I addressed the review comments, fixed `getUnsubscribeUrl` to work with `subscribers` and `email_addresses` tables, added a test for `getEmailAddressesByHashes`, and fixed `lint` things.

This PR is starting to scope-creep, so I'd like to tie this off and merge it into `monitor_v2` sooner, and start moar/new branches to do things like fix tests for v2.

pdehaan commented:
Looks like some test failures in https://travis-ci.org/mozilla/blurts-server/jobs/529013678

groovecoder commented:
I have another branch - [`fix-moar-v2-tests`](https://github.com/mozilla/blurts-server/tree/fix-moar-v2-tests) - that starts fixing most/all tests. I rebased it on this branch and plan to use it to fix all the tests at once before merging `monitor_v2` into `master`.
